File: controllers/expenses.js
const { Expense, ExpenseCategory, AppUser, Household, HouseholdMember, Card, RecurringExpense, TransactionAttachment, sequelize } = require('../models');
const { Op } = require('sequelize');
const { deleteAttachmentFilesForTransactions } = require('./attachments');
const {
  getCurrentFinancialPeriod,
  getFinancialPeriod,
  getCalendarMonthPeriod,
  getFinancialWeeks,
  getHourlySlots,
  getDailySlots
} = require('../utils/financialCalendar');

/**
 * Create one Expense row (or a chain of installment rows).
 * Shared by the HTTP `create` handler and by shopping-session completion.
 *
 * Returns the first Expense (the head of the installment chain), reloaded
 * with ExpenseCategory + AppUser includes.
 */
async function createExpenseRecord({
  amount,
  dateTime,
  description,
  note,
  paymentMethod,
  cardId,
  expenseCategoryId,
  householdId,
  appUserId,
  installmentTotal,
  installmentCurrent
}) {
  const total   = installmentTotal && Number(installmentTotal) > 1 ? Number(installmentTotal) : null;
  const current = total ? (installmentCurrent ? Number(installmentCurrent) : 1) : null;

  let firstExpense;
  let parentId = null;
  if (total && total > 1) {
    const baseDate = new Date(dateTime);
    for (let i = 0; i < total; i++) {
      const paymentDate = new Date(baseDate);
      paymentDate.setMonth(paymentDate.getMonth() + i);
      const rec = await Expense.create({
        amount,
        dateTime:           paymentDate.toISOString(),
        description:        description || null,
        note:               note        || null,
        paymentMethod,
        cardId:             cardId ? Number(cardId) : null,
        expenseCategoryId:  Number(expenseCategoryId),
        appUserId,
        householdId:        Number(householdId),
        installmentTotal:   total,
        installmentCurrent: current + i,
        parentExpenseId:    i === 0 ? null : parentId
      });
      if (i === 0) {
        firstExpense = rec;
        parentId = rec.id;
      }
    }
  } else {
    firstExpense = await Expense.create({
      amount,
      dateTime,
      description: description || null,
      note:        note        || null,
      paymentMethod,
      cardId:      cardId ? Number(cardId) : null,
      expenseCategoryId: Number(expenseCategoryId),
      appUserId,
      householdId: Number(householdId),
      installmentTotal:  null,
      installmentCurrent: null
    });
  }

  return Expense.findByPk(firstExpense.id, {
    include: [
      { model: ExpenseCategory, attributes: ['id', 'name', 'nameHe', 'icon', 'color'] },
      { model: AppUser,         attributes: ['id', 'username'] },
      {
        model:    TransactionAttachment,
        as:       'attachments',
        attributes: ['id', 'filename', 'originalFilename', 'mimeType', 'size', 'isImage', 'createdAt'],
        required: false
      }
    ]
  });
}

/**
 * Given a recurring expense's dayOfMonth and the full financial period,
 * returns the Date when it occurs within that period, or null if it doesn't.
 */
function getRecurringOccurrence(dayOfMonth, period) {
  // Try the first calendar month of the period (period.start month)
  const startYear  = period.start.getFullYear();
  const startMonth = period.start.getMonth();
  const d1 = new Date(startYear, startMonth, dayOfMonth, 12, 0, 0, 0);
  if (d1 >= period.start && d1 <= period.end) return d1;

  // Try the second calendar month of the period (period.end month)
  const endYear  = period.end.getFullYear();
  const endMonth = period.end.getMonth();
  const d2 = new Date(endYear, endMonth, dayOfMonth, 12, 0, 0, 0);
  if (d2 >= period.start && d2 <= period.end) return d2;

  return null;
}

class ExpensesController {
  /**
   * GET /api/app/expenses
   * List expenses for a household with financial-calendar scoping.
   *
   * Query params:
   *   householdId   (required)
   *   view          'hourly' | 'daily' | 'weekly' | 'monthly'  (default 'monthly')
   *   periodOffset  integer  (default 0)
   *   weekNumber    1-5  (required for hourly / daily views)
   *   date          ISO string  (required for hourly view)
   *   categoryId    optional filter
   */
  async list(ctx) {
    try {
      const {
        householdId,
        view          = 'monthly',
        periodOffset  = 0,
        periodType    = 'financial',
        weekNumber,
        date,
        categoryId
      } = ctx.request.query;

      // --- Validate householdId ---
      if (!householdId) {
        ctx.status = 400;
        ctx.body   = { error: 'householdId is required' };
        return;
      }

      // --- Verify user is a member of the household ---
      const appUserId  = ctx.state.appUser.id;
      const membership = await HouseholdMember.findOne({
        where: { householdId: Number(householdId), appUserId }
      });

      if (!membership) {
        ctx.status = 403;
        ctx.body   = { error: 'You are not a member of this household' };
        return;
      }

      // --- Resolve the financial period ---
      const offset = parseInt(periodOffset, 10) || 0;
      const household = await Household.findByPk(Number(householdId), {
        attributes: ['id', 'financialMonthStartDay']
      });
      const startDay = household?.financialMonthStartDay ?? 10;
      const period = periodType === 'calendar'
        ? getCalendarMonthPeriod(offset)
        : getFinancialPeriod(offset, startDay);
      const weeks  = getFinancialWeeks(period);

      // --- Build date range based on view ---
      let dateStart = period.start;
      let dateEnd   = period.end;
      let slots     = null;

      if (view === 'weekly') {
        // Use the full financial period; no weekNumber required for this view
        // dateStart and dateEnd already set to period.start / period.end

      } else if (view === 'daily') {
        if (!date) {
          ctx.status = 400;
          ctx.body   = { error: 'date is required for daily view' };
          return;
        }
        const targetDate = new Date(date);
        if (isNaN(targetDate.getTime())) {
          ctx.status = 400;
          ctx.body   = { error: 'date is not a valid ISO date string' };
          return;
        }
        dateStart = new Date(targetDate.getFullYear(), targetDate.getMonth(), targetDate.getDate(), 0, 0, 0, 0);
        dateEnd   = new Date(targetDate.getFullYear(), targetDate.getMonth(), targetDate.getDate(), 23, 59, 59, 999);

      } else if (view === 'hourly') {
        if (!date) {
          ctx.status = 400;
          ctx.body   = { error: 'date is required for hourly view' };
          return;
        }
        const targetDate = new Date(date);
        if (isNaN(targetDate.getTime())) {
          ctx.status = 400;
          ctx.body   = { error: 'date is not a valid ISO date string' };
          return;
        }
        dateStart = new Date(targetDate.getFullYear(), targetDate.getMonth(), targetDate.getDate(), 0, 0, 0, 0);
        dateEnd   = new Date(targetDate.getFullYear(), targetDate.getMonth(), targetDate.getDate(), 23, 59, 59, 999);
        slots     = getHourlySlots(targetDate);
      }

      // --- Build where clause ---
      const where = {
        householdId: Number(householdId),
        dateTime: {
          [Op.between]: [dateStart, dateEnd]
        }
      };

      if (categoryId) {
        where.expenseCategoryId = Number(categoryId);
      }

      // --- Fetch regular expenses ---
      const expenses = await Expense.findAll({
        where,
        include: [
          {
            model:      ExpenseCategory,
            attributes: ['id', 'name', 'nameHe', 'icon', 'color']
          },
          {
            model:      AppUser,
            attributes: ['id', 'username']
          },
          {
            model:      Card,
            as:         'card',
            attributes: ['id', 'lastFourDigits', 'nickname', 'bankName', 'cardType'],
            required:   false
          },
          {
            model:      TransactionAttachment,
            as:         'attachments',
            attributes: ['id', 'filename', 'originalFilename', 'mimeType', 'size', 'isImage', 'createdAt'],
            required:   false
          }
        ],
        order: [['dateTime', 'DESC']]
      });

      // --- Fetch and merge recurring expenses ---
      const allRecurring = await RecurringExpense.findAll({
        where: { householdId: Number(householdId), isActive: true },
        include: [
          { model: ExpenseCategory, attributes: ['id', 'name', 'nameHe', 'icon', 'color'] },
          { model: AppUser,         attributes: ['id', 'username'] }
        ]
      });

      const recurringEntries = [];
      for (const rec of allRecurring) {
        const occurrenceDate = getRecurringOccurrence(rec.dayOfMonth, period);
        if (!occurrenceDate) continue;

        // Check it falls within the (potentially narrower) sub-period window
        if (occurrenceDate < dateStart || occurrenceDate > dateEnd) continue;

        // Check startYear/startMonth <= occurrence calendar month
        const occYear  = occurrenceDate.getFullYear();
        const occMonth = occurrenceDate.getMonth() + 1; // 1-based
        if (rec.startYear > occYear || (rec.startYear === occYear && rec.startMonth > occMonth)) continue;

        const r = rec.toJSON();
        recurringEntries.push({
          ...r,
          dateTime:           occurrenceDate.toISOString(),
          isRecurring:        true,
          recurringExpenseId: r.id,
          // Reshape associations to match Expense format expected by Flutter
          ExpenseCategory: r.ExpenseCategory,
          AppUser:         r.AppUser,
          card:            null,
          cardId:          null,
          installmentTotal:   null,
          installmentCurrent: null,
          parentExpenseId:    null
        });
      }

      const baseUrl = `${ctx.protocol}://${ctx.host}`;

      function enrichAttachments(expenseObj) {
        if (!Array.isArray(expenseObj.attachments)) {
          expenseObj.attachments = [];
        }
        expenseObj.attachments = expenseObj.attachments.map(a => ({
          ...a,
          fileUrl:  `${baseUrl}/api/app/attachments/${a.id}/file`,
          thumbUrl: a.isImage ? `${baseUrl}/api/app/attachments/${a.id}/thumb` : null
        }));
        expenseObj.attachmentCount = expenseObj.attachments.length;
        return expenseObj;
      }

      const allExpenses = [
        ...expenses.map(e => enrichAttachments(e.toJSON())),
        ...recurringEntries.map(r => ({ ...r, attachments: [], attachmentCount: 0 }))
      ].sort((a, b) => new Date(b.dateTime) - new Date(a.dateTime));

      // --- Compute total ---
      const totalAmount = allExpenses.reduce((sum, e) => sum + parseFloat(e.amount), 0);

      // --- Build response ---
      const response = {
        success:     true,
        expenses:    allExpenses,
        period:      { start: period.start, end: period.end, label: period.label },
        totalAmount: Math.round(totalAmount * 100) / 100
      };

      if (view === 'monthly' || view === 'weekly') {
        response.weeks = weeks;
      }

      if (slots) {
        response.slots = slots;
      }

      ctx.body = response;

    } catch (error) {
      console.error('Expenses list error:', error);
      ctx.status = 500;
      ctx.body   = { error: 'Failed to fetch expenses' };
    }
  }

  /**
   * POST /api/app/expenses
   * Create a new expense transaction.
   *
   * Body: { amount, dateTime, description, note, paymentMethod, expenseCategoryId, householdId }
   */
  async create(ctx) {
    try {
      const {
        amount,
        dateTime,
        description,
        note,
        paymentMethod,
        cardId,
        expenseCategoryId,
        householdId,
        installmentTotal,
        installmentCurrent
      } = ctx.request.body;

      // --- Required fields ---
      if (!amount || !dateTime || !expenseCategoryId || !householdId) {
        ctx.status = 400;
        ctx.body   = { error: 'amount, dateTime, expenseCategoryId, and householdId are required' };
        return;
      }

      // --- Verify user is a member of the household ---
      const appUserId  = ctx.state.appUser.id;
      const membership = await HouseholdMember.findOne({
        where: { householdId: Number(householdId), appUserId }
      });

      if (!membership) {
        ctx.status = 403;
        ctx.body   = { error: 'You are not a member of this household' };
        return;
      }

      // --- Validate category belongs to household ---
      const category = await ExpenseCategory.findOne({
        where: { id: Number(expenseCategoryId), householdId: Number(householdId) }
      });

      if (!category) {
        ctx.status = 400;
        ctx.body   = { error: 'Invalid expenseCategoryId for this household' };
        return;
      }

      const created = await createExpenseRecord({
        amount,
        dateTime,
        description,
        note,
        paymentMethod,
        cardId,
        expenseCategoryId,
        householdId,
        appUserId,
        installmentTotal,
        installmentCurrent
      });

      const baseUrl = `${ctx.protocol}://${ctx.host}`;
      const expenseJson = created.toJSON();
      if (!Array.isArray(expenseJson.attachments)) expenseJson.attachments = [];
      expenseJson.attachments = expenseJson.attachments.map(a => ({
        ...a,
        fileUrl:  `${baseUrl}/api/app/attachments/${a.id}/file`,
        thumbUrl: a.isImage ? `${baseUrl}/api/app/attachments/${a.id}/thumb` : null
      }));
      expenseJson.attachmentCount = expenseJson.attachments.length;

      ctx.status = 201;
      ctx.body   = { success: true, expense: expenseJson };

    } catch (error) {
      console.error('Expenses create error:', error);
      ctx.status = 500;
      ctx.body   = { error: 'Failed to create expense' };
    }
  }

  /**
   * PUT /api/app/expenses/:id
   * Update an existing expense (only the creator can update).
   *
   * Body: { amount, dateTime, description, note, paymentMethod, expenseCategoryId }
   */
  async update(ctx) {
    try {
      const { id }   = ctx.params;
      const appUserId = ctx.state.appUser.id;

      const expense = await Expense.findByPk(id);

      if (!expense) {
        ctx.status = 404;
        ctx.body   = { error: 'Expense not found' };
        return;
      }

      if (expense.appUserId !== appUserId) {
        ctx.status = 403;
        ctx.body   = { error: 'You can only update your own expenses' };
        return;
      }

      const {
        amount,
        dateTime,
        description,
        note,
        paymentMethod,
        cardId,
        expenseCategoryId,
        installmentTotal,
        installmentCurrent
      } = ctx.request.body;

      // --- Validate new category if provided ---
      if (expenseCategoryId) {
        const category = await ExpenseCategory.findOne({
          where: { id: Number(expenseCategoryId), householdId: expense.householdId }
        });

        if (!category) {
          ctx.status = 400;
          ctx.body   = { error: 'Invalid expenseCategoryId for this household' };
          return;
        }
      }

      // --- Apply updates ---
      await expense.update({
        amount:             amount             !== undefined ? amount             : expense.amount,
        dateTime:           dateTime           !== undefined ? dateTime           : expense.dateTime,
        description:        description        !== undefined ? description        : expense.description,
        note:               note               !== undefined ? note               : expense.note,
        paymentMethod:      paymentMethod      !== undefined ? paymentMethod      : expense.paymentMethod,
        cardId:             cardId             !== undefined ? (cardId ? Number(cardId) : null) : expense.cardId,
        expenseCategoryId:  expenseCategoryId  !== undefined ? Number(expenseCategoryId) : expense.expenseCategoryId,
        installmentTotal:   installmentTotal   !== undefined ? (installmentTotal ? Number(installmentTotal) : null) : expense.installmentTotal,
        installmentCurrent: installmentCurrent !== undefined ? (installmentCurrent ? Number(installmentCurrent) : null) : expense.installmentCurrent
      });

      // --- Reload with includes ---
      const updated = await Expense.findByPk(expense.id, {
        include: [
          {
            model:      ExpenseCategory,
            attributes: ['id', 'name', 'nameHe', 'icon', 'color']
          },
          {
            model:      AppUser,
            attributes: ['id', 'username']
          }
        ]
      });

      ctx.body = { success: true, expense: updated.toJSON() };

    } catch (error) {
      console.error('Expenses update error:', error);
      ctx.status = 500;
      ctx.body   = { error: 'Failed to update expense' };
    }
  }

  /**
   * PUT /api/app/expenses/:id/installment-amount
   * Update the amount of an installment expense across a chosen scope.
   *
   * Body: { amount, scope }  — scope: 'all' | 'forward' | 'this'
   *   all     → update every record in the installment group
   *   forward → update this record and every later installment (installmentCurrent >=)
   *   this    → update only this record
   */
  async updateInstallmentAmount(ctx) {
    try {
      const { id }    = ctx.params;
      const appUserId = ctx.state.appUser.id;
      const { amount, scope } = ctx.request.body;

      if (amount === undefined || !scope || !['all', 'forward', 'this'].includes(scope)) {
        ctx.status = 400;
        ctx.body   = { error: 'amount and scope (all|forward|this) are required' };
        return;
      }

      const expense = await Expense.findByPk(id);
      if (!expense) {
        ctx.status = 404;
        ctx.body   = { error: 'Expense not found' };
        return;
      }

      if (expense.appUserId !== appUserId) {
        ctx.status = 403;
        ctx.body   = { error: 'You can only update your own expenses' };
        return;
      }

      if (scope === 'this') {
        await expense.update({ amount });
        ctx.body = { success: true };
        return;
      }

      // Determine the root of this installment group
      const rootId = expense.parentExpenseId ?? expense.id;

      const where = {
        [Op.or]: [{ id: rootId }, { parentExpenseId: rootId }]
      };

      if (scope === 'forward') {
        where.installmentCurrent = { [Op.gte]: expense.installmentCurrent };
      }

      await Expense.update({ amount }, { where });

      ctx.body = { success: true };
    } catch (error) {
      console.error('Expenses updateInstallmentAmount error:', error);
      ctx.status = 500;
      ctx.body   = { error: 'Failed to update installment amounts' };
    }
  }

  /**
   * DELETE /api/app/expenses/:id
   * Delete an expense (only the creator can delete).
   */
  async delete(ctx) {
    try {
      const { id }    = ctx.params;
      const appUserId  = ctx.state.appUser.id;

      const expense = await Expense.findByPk(id);

      if (!expense) {
        ctx.status = 404;
        ctx.body   = { error: 'Expense not found' };
        return;
      }

      if (expense.appUserId !== appUserId) {
        ctx.status = 403;
        ctx.body   = { error: 'You can only delete your own expenses' };
        return;
      }

      // Collect all expense IDs in this chain (this + children)
      const childExpenses = await Expense.findAll({
        where:      { parentExpenseId: expense.id },
        attributes: ['id']
      });
      const allExpenseIds = [expense.id, ...childExpenses.map(c => c.id)];

      // Remove attachment files from disk for all expenses in this chain
      // (DB rows will be removed by ON DELETE CASCADE after expense rows are destroyed)
      await deleteAttachmentFilesForTransactions({ expenseIds: allExpenseIds });

      // Cascade-delete all future installments linked to this expense
      await Expense.destroy({ where: { parentExpenseId: expense.id } });
      await expense.destroy();

      ctx.body = { success: true };

    } catch (error) {
      console.error('Expenses delete error:', error);
      ctx.status = 500;
      ctx.body   = { error: 'Failed to delete expense' };
    }
  }
}

module.exports = new ExpensesController();
module.exports.createExpenseRecord = createExpenseRecord;

File: backend_go/internal/handlers/expenses.go
package handlers

import (
	"household-go/internal/database"
	"household-go/internal/models"
	"household-go/internal/utils"
	"log"
	"math"
	"time"

	"github.com/gin-gonic/gin"
	"gorm.io/gorm"
)

// ExpensesHandler handles all expense CRUD and listing.
type ExpensesHandler struct{}

func NewExpensesHandler() *ExpensesHandler { return &ExpensesHandler{} }

// List returns expenses for a household scoped by the financial calendar.
// GET /api/app/expenses?householdId=X&view=monthly&periodOffset=0
func (h *ExpensesHandler) List(c *gin.Context) {
	appUser := c.MustGet("appUser").(*models.AppUser)

	householdID := c.Query("householdId")
	if householdID == "" {
		utils.BadRequest(c, "householdId is required")
		return
	}

	// Verify membership — identical to the Sequelize HouseholdMember.findOne check
	var membership models.HouseholdMember
	if err := database.DB.Where("household_id = ? AND app_user_id = ?", householdID, appUser.ID).
		First(&membership).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			utils.Forbidden(c, "You are not a member of this household")
		} else {
			utils.InternalError(c, "Failed to fetch expenses")
		}
		return
	}

	view := c.DefaultQuery("view", "monthly")
	periodOffset := parseInt(c.DefaultQuery("periodOffset", "0"))
	dateStr := c.Query("date")
	categoryID := c.Query("categoryId")

	period := utils.GetFinancialPeriod(periodOffset)
	weeks := utils.GetFinancialWeeks(period)

	dateStart := period.Start
	dateEnd := period.End
	var slots interface{} // will hold []HourlySlot or nil

	switch view {
	case "daily":
		if dateStr == "" {
			utils.BadRequest(c, "date is required for daily view")
			return
		}
		t, err := time.Parse(time.RFC3339, dateStr)
		if err != nil {
			// Try date-only format as fallback
			t, err = time.Parse("2006-01-02", dateStr)
			if err != nil {
				utils.BadRequest(c, "date is not a valid ISO date string")
				return
			}
		}
		dateStart = time.Date(t.Year(), t.Month(), t.Day(), 0, 0, 0, 0, time.Local)
		dateEnd = time.Date(t.Year(), t.Month(), t.Day(), 23, 59, 59, 0, time.Local)

	case "hourly":
		if dateStr == "" {
			utils.BadRequest(c, "date is required for hourly view")
			return
		}
		t, err := time.Parse(time.RFC3339, dateStr)
		if err != nil {
			t, err = time.Parse("2006-01-02", dateStr)
			if err != nil {
				utils.BadRequest(c, "date is not a valid ISO date string")
				return
			}
		}
		dateStart = time.Date(t.Year(), t.Month(), t.Day(), 0, 0, 0, 0, time.Local)
		dateEnd = time.Date(t.Year(), t.Month(), t.Day(), 23, 59, 59, 0, time.Local)
		slots = utils.GetHourlySlots(t)
	}

	// Build the GORM query.
	// Preload() is GORM's equivalent of Sequelize's `include: [{ model: ExpenseCategory }]`.
	// It runs a separate SELECT and joins the results in Go (not a SQL JOIN).
	// For a household-scale app this is fine; for high-traffic use .Joins() instead.
	query := database.DB.
		Where("household_id = ?", householdID).
		Where("date_time BETWEEN ? AND ?", dateStart, dateEnd).
		Preload("ExpenseCategory", func(db *gorm.DB) *gorm.DB {
			return db.Select("id, name, name_he, icon, color")
		}).
		Preload("AppUser", func(db *gorm.DB) *gorm.DB {
			return db.Select("id, username")
		}).
		Order("date_time DESC")

	if categoryID != "" {
		query = query.Where("expense_category_id = ?", categoryID)
	}

	var expenses []models.Expense
	if err := query.Find(&expenses).Error; err != nil {
		log.Printf("Expenses list error: %v", err)
		utils.InternalError(c, "Failed to fetch expenses")
		return
	}

	// Compute total — Go's range loop iterates slices with index, value
	total := 0.0
	for _, e := range expenses {
		total += e.Amount
	}
	// Round to 2 decimal places (avoid floating-point drift)
	total = math.Round(total*100) / 100

	resp := gin.H{
		"success":     true,
		"expenses":    expenses,
		"period":      gin.H{"start": period.Start, "end": period.End, "label": period.Label},
		"totalAmount": total,
	}

	if view == "monthly" || view == "weekly" {
		resp["weeks"] = weeks
	}
	if slots != nil {
		resp["slots"] = slots
	}

	utils.OK(c, resp)
}

// Create adds a new expense.
// POST /api/app/expenses
func (h *ExpensesHandler) Create(c *gin.Context) {
	appUser := c.MustGet("appUser").(*models.AppUser)

	var body struct {
		Amount            float64  `json:"amount"            binding:"required"`
		DateTime          string   `json:"dateTime"          binding:"required"`
		Description       *string  `json:"description"`
		Note              *string  `json:"note"`
		PaymentMethod     *string  `json:"paymentMethod"`
		ExpenseCategoryID uint     `json:"expenseCategoryId" binding:"required"`
		HouseholdID       uint     `json:"householdId"       binding:"required"`
	}
	if err := c.ShouldBindJSON(&body); err != nil {
		utils.BadRequest(c, "amount, dateTime, expenseCategoryId, and householdId are required")
		return
	}

	// Membership check
	var membership models.HouseholdMember
	if err := database.DB.Where("household_id = ? AND app_user_id = ?", body.HouseholdID, appUser.ID).
		First(&membership).Error; err != nil {
		utils.Forbidden(c, "You are not a member of this household")
		return
	}

	// Validate category belongs to this household
	var category models.ExpenseCategory
	if err := database.DB.Where("id = ? AND household_id = ?", body.ExpenseCategoryID, body.HouseholdID).
		First(&category).Error; err != nil {
		utils.BadRequest(c, "Invalid expenseCategoryId for this household")
		return
	}

	// Parse dateTime — try RFC3339 first, then date-only
	dt, err := parseDateTime(body.DateTime)
	if err != nil {
		utils.BadRequest(c, "Invalid dateTime format")
		return
	}

	expense := models.Expense{
		Amount:            body.Amount,
		DateTime:          dt,
		Description:       body.Description,
		Note:              body.Note,
		PaymentMethod:     body.PaymentMethod,
		ExpenseCategoryID: body.ExpenseCategoryID,
		AppUserID:         appUser.ID,
		HouseholdID:       body.HouseholdID,
	}

	if err := database.DB.Create(&expense).Error; err != nil {
		log.Printf("Expenses create error: %v", err)
		utils.InternalError(c, "Failed to create expense")
		return
	}

	// Reload with associations — equivalent to Sequelize's findByPk with include
	database.DB.Preload("ExpenseCategory", func(db *gorm.DB) *gorm.DB {
		return db.Select("id, name, name_he, icon, color")
	}).Preload("AppUser", func(db *gorm.DB) *gorm.DB {
		return db.Select("id, username")
	}).First(&expense, expense.ID)

	utils.Created(c, gin.H{"success": true, "expense": expense})
}

// Update modifies an existing expense (only the creator can update).
// PUT /api/app/expenses/:id
func (h *ExpensesHandler) Update(c *gin.Context) {
	appUser := c.MustGet("appUser").(*models.AppUser)
	id := c.Param("id")

	var expense models.Expense
	if err := database.DB.First(&expense, id).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			utils.NotFound(c, "Expense not found")
		} else {
			utils.InternalError(c, "Failed to update expense")
		}
		return
	}

	if expense.AppUserID != appUser.ID {
		utils.Forbidden(c, "You can only update your own expenses")
		return
	}

	// Use a map so we can do partial updates without overwriting fields with zero values.
	// In Go, a struct update with Updates() skips zero-value fields — problematic if
	// a user legitimately wants to set amount to 0. A map avoids this.
	var body struct {
		Amount            *float64 `json:"amount"`
		DateTime          *string  `json:"dateTime"`
		Description       *string  `json:"description"`
		Note              *string  `json:"note"`
		PaymentMethod     *string  `json:"paymentMethod"`
		ExpenseCategoryID *uint    `json:"expenseCategoryId"`
	}
	if err := c.ShouldBindJSON(&body); err != nil {
		utils.BadRequest(c, "Invalid request body")
		return
	}

	updates := map[string]interface{}{}
	if body.Amount != nil {
		updates["amount"] = *body.Amount
	}
	if body.DateTime != nil {
		dt, err := parseDateTime(*body.DateTime)
		if err != nil {
			utils.BadRequest(c, "Invalid dateTime format")
			return
		}
		updates["date_time"] = dt
	}
	if body.Description != nil {
		updates["description"] = body.Description
	}
	if body.Note != nil {
		updates["note"] = body.Note
	}
	if body.PaymentMethod != nil {
		updates["payment_method"] = body.PaymentMethod
	}
	if body.ExpenseCategoryID != nil {
		// Validate the new category belongs to the same household
		var cat models.ExpenseCategory
		if err := database.DB.Where("id = ? AND household_id = ?", *body.ExpenseCategoryID, expense.HouseholdID).
			First(&cat).Error; err != nil {
			utils.BadRequest(c, "Invalid expenseCategoryId for this household")
			return
		}
		updates["expense_category_id"] = *body.ExpenseCategoryID
	}

	if len(updates) > 0 {
		database.DB.Model(&expense).Updates(updates)
	}

	database.DB.Preload("ExpenseCategory", func(db *gorm.DB) *gorm.DB {
		return db.Select("id, name, name_he, icon, color")
	}).Preload("AppUser", func(db *gorm.DB) *gorm.DB {
		return db.Select("id, username")
	}).First(&expense, expense.ID)

	utils.OK(c, gin.H{"success": true, "expense": expense})
}

// Delete removes an expense (only creator can delete).
// DELETE /api/app/expenses/:id
func (h *ExpensesHandler) Delete(c *gin.Context) {
	appUser := c.MustGet("appUser").(*models.AppUser)
	id := c.Param("id")

	var expense models.Expense
	if err := database.DB.First(&expense, id).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			utils.NotFound(c, "Expense not found")
		} else {
			utils.InternalError(c, "Failed to delete expense")
		}
		return
	}

	if expense.AppUserID != appUser.ID {
		utils.Forbidden(c, "You can only delete your own expenses")
		return
	}

	database.DB.Delete(&expense)
	utils.OK(c, gin.H{"success": true})
}

// parseInt is a helper to parse a string to int, returning 0 on failure.
// Go has no parseInt() builtin — you use strconv.Atoi() which returns (int, error).
// The underscore _ discards the error (intentional fallback to 0).
func parseInt(s string) int {
	n := 0
	for _, ch := range s {
		if ch < '0' || ch > '9' {
			if ch == '-' && n == 0 {
				n = -1
				continue
			}
			break
		}
		n = n*10 + int(ch-'0')
	}
	return n
}

// parseDateTime tries RFC3339 ("2024-03-15T10:00:00Z") then date-only ("2024-03-15").
func parseDateTime(s string) (time.Time, error) {
	if t, err := time.Parse(time.RFC3339, s); err == nil {
		return t, nil
	}
	return time.Parse("2006-01-02", s)
}
